plugins/catalog-backend-module-backstage/src/provider/BackstageEntityProvider.ts

```diff
@@ -0,0 +1,354 @@
+  }
+
+  private createScheduleFn(taskRunner: TaskRunner): () => Promise<void> {
+    return async () => {
+      const taskId = `${this.getProviderName()}:refresh`;
+      return taskRunner.run({
+        id: taskId,
+        fn: async () => {
+          const logger = this.logger.child({
+            class: BackstageEntityProvider.prototype.constructor.name,
+            taskId,
+            taskInstanceId: uuid.v4(),
+          });
+          try {
+            await this.refresh(logger);
+          } catch (error) {
+            logger.error(
+              `${this.getProviderName()} refresh failed, ${error}`,
+              error,
+            );
+          }
+        },
+      });
+    };
+  }
+
+  async refresh(logger: Logger) {
+    if (!this.connection) {
+      throw new Error('Not initialized');
+    }
+    const enabled = this.config.getBoolean(
+      'catalog.providers.backstage.enabled',
+    );
+    if (!enabled) {
+      logger.warn(
+        'Backstage entity provider is currently disabled via config, exiting',
+      );
+      return;
+    }
+
+    const entities: Entity[] = [];
+
+    const repositories = await this.getAllRepositories();
+
+    for (const repository of repositories) {
+      const readTreeResponse = await this.urlReader.readTree(
+        `https://github.com/backstage/${repository.name}`,
+      );
+
+      const fileDir = await readTreeResponse.dir();
+      const catalogInfoFiles = globSync(
+        [
+          'catalog-info.yaml',
+          'packages/*/catalog-info.yaml',
+          'plugins/*/catalog-info.yaml',
+        ],
+        { cwd: fileDir },
+      );
+
+      logger.info(
+        `Found ${catalogInfoFiles.length} catalog-info.yaml files in the '${repository.name}' repository`,
+      );
+
+      if (!catalogInfoFiles) {
+        // No catalog info files found, continue
+        logger.warn(
+          `No catalog-info.yaml file found in the '${repository.name}' repository, continuing`,
+        );
+        continue;
+      }
+
+      for (const catalogInfoFile of catalogInfoFiles) {
+        const catalogInfoDir = path.join(fileDir, catalogInfoFile);
+        const rawEntities = await this.parseEntityYaml(catalogInfoDir);
+        const cleanEntities = this.cleanUpRawEntities(
+          rawEntities,
+          repository,
+          catalogInfoFile,
+        );
+
+        const mappedEntities: Entity[] = [];
+        for (const cleanEntity of cleanEntities) {
+          switch (cleanEntity.kind) {
+            case 'Component': {
+              const mappedComponent = await this.componentMapper(
+                cleanEntity,
+                repository,
+                catalogInfoDir,
+              );
+              mappedEntities.push(mappedComponent);
+              break;
+            }
+
+            default:
+              mappedEntities.push(cleanEntity);
+          }
+        }
+
+        entities.push(...mappedEntities);
+      }
+
+      await fs.remove(fileDir);
```

Comment:
bonus points to do `try { ... } finally { fs.remove }` to ensure cleanup

Reply:
Yes, much nicer. 👍